src/dcegm/process_model.py

```diff
@@ -0,0 +1,372 @@
+
+
+def get_exog_transition_vec(state_choice_vec, exog_mapping, exog_funcs, params):
+    exog_state_global = state_choice_vec[-2]
+
+    _idx_endog_states_and_choice = list(range(len(state_choice_vec) - 2)) + [-1]
+    endog_states_and_choice = state_choice_vec[jnp.array(_idx_endog_states_and_choice)]
+
+    trans_vecs = []
+
+    for exog_state, exog_func in zip(exog_mapping[exog_state_global], exog_funcs):
+        # options already partialled in
+        trans_vec = exog_func(exog_state, *endog_states_and_choice, params)
+        trans_vecs.append(jnp.array(trans_vec))
+
+    trans_vec_kron = reduce(jnp.kron, trans_vecs)
+
+    return jnp.array(trans_vec_kron)
+
+
+def create_exog_mapping(options):
+    """Create mapping from separate exog state variables to global exog state."""
+
+    exog_state_vars = options["state_variables"]["exogenous"]
+
+    n_elements = []
+    for key in exog_state_vars:
+        n_elements.append(len(exog_state_vars[key]))
+
+    def recursive_generator(n_elements, current_mapping=[]):
+        if not n_elements:
+            exog_mapping.append(current_mapping)
+            return
+
+        for i in range(n_elements[0]):
+            recursive_generator(n_elements[1:], current_mapping + [i])
+
+    exog_mapping = []
+    recursive_generator(n_elements)
+
+    return jnp.asarray(exog_mapping)
+
+
+def _get_exog_function_with_filtered_args(func, options):
+    """Args is state_vec with one global exog state."""
+    signature = list(inspect.signature(func).parameters)
+    exog_key = signature & options["state_variables"]["exogenous"].keys()
+
+    _endog_state_vars_and_choice = options["state_variables"]["endogenous"] | {
+        "choice": options["state_variables"]["choice"]
+    }
+    endog_to_index = {
+        key: idx for idx, key in enumerate(_endog_state_vars_and_choice.keys())
+    }
+
+    @functools.wraps(func)
+    def processed_func(*args):
+        exog_arg, endog_args = args[0], args[1:-1]
+
+        # Allows for flexible position of arguments in user function.
+        # The endog states and choice variable in the state_choice_vec
+        # (passed as args to the user function) have the same order
+        # as they appear in options["state_variables"]
+        _args_to_kwargs = {
+            key: endog_args[idx]
+            for key, idx in endog_to_index.items()
+            if key in signature  # and key != "choice"
+        }
+
+        if exog_key:
+            _args_to_kwargs[exog_key.pop()] = exog_arg
+
+        if "params" in signature:
+            _args_to_kwargs["params"] = args[-1]
+
+        # partial in
+        if "options" in signature and "model_params" in options:
+            _args_to_kwargs["options"] = options["model_params"]
+
+        return func(**_args_to_kwargs)
+
+    # Set name of the original func
+    processed_func.__name__ = func.__name__
+
+    return processed_func
+
+
+def _get_utility_function_with_filtered_args_and_kwargs(func, options, exog_mapping):
```

Comment:
Lets zoom tomorrow and maybe you can walk me through that function. I think we can simplify all of that, but we can do that in a different PR and I can do it. To better understand!